### DISCUSSION TITLE
override buildtime with runtime env variables?

### DISCUSSION THREAD

**Discussion by lifeofmoo** [vercel/next.js#43721] [Help]
Hello,

Is there a way to override NextJS buildtime variables with runtime vars in the server.js file?

Our issue:

We build NextJS docker images via Github Actions and deploy to K8s (AWS EKS)

A snippet of our docker file is below:

```
COPY package.json package-lock.json ./ 
RUN npm install

FROM node:16-alpine AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

RUN npm run build

FROM node:16-alpine AS runner
WORKDIR /app

ENV NODE_ENV production

...

EXPOSE 8080

ENV PORT 8080

CMD ["node", "server.js"]
```

In the root of the repo we have an **.env.production** file with one environment variable:

`DEV_API=https://backend-api.dev.eks.xxxxxx.com
`

The Github action builds and pushes this image to our container repo and when it runs in EKS we have this in the server.js file.

```
/app $ cat server.js 

process.env.NODE_ENV = 'production'
process.chdir(__dirname)
const NextServer = require('next/dist/server/next-server').default
const http = require('http')
const path = require('path')

...

server.listen(currentPort, (err) => {
  if (err) {
    console.error("Failed to start server", err)
    process.exit(1)
  }
  const nextServer = new NextServer({
    hostname: 'localhost',
    port: currentPort,
    dir: path.join(__dirname),
    dev: false,
    customServer: false,
    conf: {"env":{"DEV_API":"https://backend-api.dev.eks.xxxxxx.com"},"webpack":null,"webpackDevMiddleware":null,"eslint":{"ignoreDuringBuilds":true},"typescript":{"ignoreBuildErrors":false,"tsconfigPath":"tsconfig.json"},"distDir":"./.next","cleanDistDir":true,"assetPrefix":"","configOrigin":"next.config.js","useFileSystemPublicRoutes":true,"generateEtags":true,"pageExtensions":["tsx","ts","jsx","js"],"target":"server","poweredByHeader":true,"compress":true,"anal

```

The problem is that if we want to **promote/deploy** this **same** image to our live cluster the server.js file is still pointing to our development backend URL.

Is there a way to parameterise this part of the server.js file?

```
conf: {"env":{"DEV_API":"https://backend-api.dev.eks.xxxxxx.com"}
```

In EKS I've created a ConfigMap and setup env variables which could be read by the container. The below is the output of running _env_ when i've exec'd onto the container.

```
/app $ env | grep dev
DEV_API=https://backend-api.dev.eks.xxxxx.com
```
I can set this to anything I want but how do I get the server.js file to take this run time variable instead of the one passed at build time (,env,production).

Our current workaround is to do something like the below, but I'd ideally I don't want to be building multiple versions of the same image just to cater for a single difference (DEV_API env variable) 

https://github.com/vercel/next.js/tree/canary/examples/with-docker-multi-env 

I've read the following articles already but no success:

https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration (only tried serverRuntimeConfig example)

https://www.twilio.com/blog/working-with-environment-variables-in-node-js-html

https://www.saltycrane.com/blog/2021/04/buildtime-vs-runtime-environment-variables-nextjs-docker/

**Comment by alicancakil** [comment]
I've been experiencing the same issue. As a team, we acknowledge **Build once, Deploy anywhere** motto.  

I want to be able to pass things in runtime, and as far as I understand, nextjs is against this idea.

Can anyone enlighten me on this subject 🙏

**Reply by emreakdas** [reply]
Have you found any solution?

**Comment by osrl** [comment]
We're using `publicRuntimeConfig` for the matter. But it has it's limitations. 

> Note: This feature is considered legacy and does not work with [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization), [Output File Tracing](https://nextjs.org/docs/advanced-features/output-file-tracing#automatically-copying-traced-files), or [React Server Components](https://nextjs.org/docs/advanced-features/react-18/server-components). Please use [environment variables](https://nextjs.org/docs/basic-features/environment-variables) instead in order to avoid initialization overhead.
